crypto/rand.go

```diff
@@ -3,32 +3,106 @@ package crypto
 import (
 	"crypto/rand"
 	"math/big"
+	"sync/atomic"
 )
 
+var (
+	// Two counters is used in case the first overflows. The first is not at
+	// risk of overflowing, but we're paranoid.
+	counter     uint64
+	counter2    uint64
+	entropyBase Hash
+)
+
+// init will create an entropy pool that the RNG will draw from. On most
+// systems, hashing out of a preset RNG pool will be substantially faster than
+// using crypto/rand, which relies on syscalls.
+func init() {
+	// Use 64 bytes in case the first 32 aren't completely random.
+	base := make([]byte, 64)
+	_, err := rand.Read(base)
+	if err != nil {
+		panic("unable to set entropy for the crypto package")
+	}
+	entropyBase = HashObject(base)
+}
+
 // RandBytes returns n bytes of random data.
 func RandBytes(n int) ([]byte, error) {
 	b := make([]byte, n)
-	_, err := rand.Read(b)
-	return b, err
+	for i := 0; i < n; i += HashSize {
+		// Fetch and update the counter values before executing the hash.
+		var newCounter, newCounter2 uint64
+		newCounter = atomic.AddUint64(&counter, 1)
+		if newCounter == 0 {
+			newCounter2 = atomic.AddUint64(&counter2, 1)
+		} else {
+			newCounter2 = atomic.LoadUint64(&counter2)
+		}
+		// Grab some entropy using the unique counter set.
+		entropy := HashAll(newCounter, newCounter2, entropyBase)
```

Comment:
The mutexes will run more noticeably slower than atomic if you have a dozen threads all concurrently calling Read I think